File: Garbage/ShadoMath/ShadoMath/Complex.cpp
#include "Complex.h"
#include <string>
#include <stdio.h>
#include <Math.h>
#include "Shado.h"


/*************************
**************************
*** Shado Complex  Lib.***
**************************
************************/

Complex::Complex(float _a, float _b) {
	a = _a;
	b = _b;
	phi = atan(b / a);
	r = sqrt(a * a + b * b);
}

Complex::Complex() {
	a = 0.0f;
	b = 0.0f;
	r = 0.0f;
	phi = 0.0f;
}

Complex::~Complex() {
}

void Complex::init(float _a, float _b) {
	a = _a;
	b = _b;
}

void Complex::initPolar(float _r, float _phi) {
	r = _r;
	phi = _phi;
	a = r * cos(phi);
	b = r * sin(phi);
}

bool Complex::equals(Complex& other) {
	if (a == other.a && b == other.b) {
		return true;
	}
	else
	{
		return false;
	}
}

std::string Complex::toString() {
	return std::to_string(this->a) + " + " + std::to_string(this->b) + " i";
}

void Complex::print() {
	std::cout << this->toString() << std::endl;
}

void Complex::printPolar() {
	printf("%1.5f * ( cos(%1.5f) + i sin(%1.5f) )", r, phi, phi);
}

Complex Complex::add(Complex& numB) const {
	return Complex(a + numB.a, b + numB.b);
}

Complex Complex::substract(Complex& numB) const {
	return Complex(a - numB.a, b - numB.b);
}

Complex Complex::multiply(Complex& numB) const {
	return Complex(a * numB.a - b * numB.b, b * numB.a + a * numB.b);
}

Complex Complex::divide(Complex& numB) const {

	float firstTerm = (a * numB.a + b * numB.b) / (numB.a * numB.a + numB.b * numB.b);
	float secondTerm = (b * numB.a - a * numB.b) / (numB.a * numB.a + numB.b * numB.b);

	return Complex(firstTerm, secondTerm);
}

float Complex::realPart() {
	return a;
}

float Complex::imaginaryPart() {
	return b;
}

float Complex::module() {
	return sqrt(a * a + b * b);
}

float Complex::angle() {
	return phi;
}

Complex Complex::conjugat() const {
	return Complex(a, -1.0f * b);
}

float Complex::mag() {
	return sqrt(a * a + b * b);
}

Complex Complex::power(int exposant) const {

	Complex* temp = new Complex();
	temp->initPolar(pow(r, exposant), exposant * phi);

	return *temp;
}

std::vector<Complex> Complex::root(int exposant) {

		std::vector<Complex> result (exposant);

		float tempR = pow(this->r, 1 / exposant);

		for (int i = 0; i < result.size(); i++) {
			float tempPhi = (phi + 2 * PI * i) / exposant;
			Complex temp = Complex();
			temp.initPolar(tempR, tempPhi);

			result[i] = temp;
		}

		return result;
}

void Complex::printSqrt(int exposant) {

	float tempR = pow(r, 1.0 / exposant);

	for (int i = 0; i < exposant; i++) {

		float tempPhi = (phi + 2 * PI * i) / exposant;

		Complex temp(0, 0);
		temp.initPolar(tempR, tempPhi);
		temp.print();
		printf("\n");
	}
}

Complex Complex::operator+(Complex other) {
	return this->add(other);
}

Complex Complex::operator-(Complex other) {
	return this->substract(other);
}

Complex Complex::operator*(Complex other) {
	return this->multiply(other);
}

Complex Complex::operator/(Complex other) {
	return this->divide(other);
}

void Complex::operator=(Complex other) {
	this->a = other.a;
	this->b = other.b;
}

bool Complex::operator==(Complex other) {
	return this->equals(other);
}

bool Complex::operator!=(Complex other) {
	return !(this->equals(other));
}

std::ostream& operator<<(std::ostream& os, Complex num)
{
	os << num.toString();
	return os;
}

File: COMP-249/Lectures/src/ShadoMath/Complex.java
/***
 *
 * Shado Complex class library
 *
 * */

package ShadoMath;

public class Complex {

	private double a = 0.0;
	private double b = 0.0;
	private double r = 0.0;
	private double phi = 0.0;

	/***
	 *
	 * @param _a The real part of the the Complex number
	 * @param _b The imaginary part of the Complex number
	 */
	public Complex(double _a, double _b) {
		this.a = _a;
		this.b = _b;
		this.r = Math.sqrt(_a * _a + _b * _b);
		this.phi = (double) Math.atan(_b / _a);
	}

	/***
	 *
	 * @param other The Complex number you want to copy
	 */
	public Complex(Complex other) {
		this(other.getA(), other.getB());
	}

	/***
	 * Default constructor initializes a to 0 and b to 0
	 */
	public Complex() {
		this(0.0, 0.0);
	}

	/***
	 * Initializes a Complex number from a polar form
	 * @param magnitude The magnitude of the Complex Number
	 * @param angle The angle (direction) of the Complex Number
	 */
	public void fromPolar(double magnitude, double angle) {
		this.r = magnitude;
		this.phi = angle;
		this.a = this.r * Math.cos(this.phi);
		this.a = this.r * Math.sin(this.phi);
	}

	/***
	 *
	 * @return Returns the magnitude of the calling Complex number
	 */
	public double mag() {
		return Math.sqrt(this.a * this.a + this.b * this.b);
	}

	/***
	 * Addes a Complex number to the calling Complex
	 * @param other The Complex you want to add
	 * @return Returns the result of the addition
	 */
	public Complex add(final Complex other) {
		return new Complex(this.a + other.getA(), this.b + other.getB());
	}

	/***
	 * Subtract a Complex number from the calling complex number																																		       a Complex number to the calling Complex
	 * @param other The Complex you want to subtract
	 * @return Returns the result of the subtraction
	 */
	public Complex subtract(final Complex other) {
		return new Complex(this.a - other.getA(), this.b - other.getB());
	}

	/***
	 * Multiply a Complex number by the calling complex number																																		       a Complex number to the calling Complex
	 * @param other The Complex you want to multiply with
	 * @return Returns the result of the multiplication
	 */
	public Complex multiply(final Complex other) {
		Complex temp = new Complex();
		temp.fromPolar(this.r * other.getR(), this.phi + other.getPhi());
		return temp;
	}

	/***
	 * Divides a Complex number by the calling complex number																																		       a Complex number to the calling Complex
	 * @param other The Complex you want to divide
	 * @return Returns the result of the division
	 */
	public Complex divide(final Complex other) {
		Complex temp = new Complex();
		temp.fromPolar(this.r / other.getR(), this.phi - other.getPhi());
		return temp;
	}

	/***
	 * Raises the calling Complex number to an exponent																																	       a Complex number to the calling Complex
	 * @param exponent The exponent
	 * @return Returns the result of the operation
	 */
	public Complex power(int exponent) {
		Complex temp = new Complex();
		temp.fromPolar(Math.pow(this.r, exponent), this.phi * exponent);
		return temp;
	}

	/***
	 * Computes the roots of the Complex number to an exponent																																	       a Complex number to the calling Complex
	 * @param exponent The strength of the root
	 * @return Returns the result of the operation
	 */
	public Complex[] root(int exponent) {

		Complex result[] = new Complex[exponent];

		double tempR = Math.pow(this.r, 1 / exponent);

		for (int i = 0; i < exponent; i++) {
			double tempPhi = (phi + 2 * Math.PI * i) / exponent;
			Complex temp = new Complex();
			temp.fromPolar(tempR, tempPhi);

			result[i] = temp;
		}

		return result;
	}

	/***
	 * Converts the current Complex number to a Vector
	 * @return Returns the resulting Vector
	 */
	public Vector toVector() {
		return new Vector(this.a, this.b);
	}

	// Setters

	/***
	 * Modifies the real part of the complex number. Recalculates the magnitude and the angle accordingly
	 * @param _a The new Real part
	 */
	public void setA(double _a) {
		this.a = _a;
		this.r = Math.sqrt(this.a * this.a + this.b * this.b);
		this.phi = Math.atan(this.b / this.a);
	}

	/***
	 * Modifies the imaginary part of the complex number. Recalculates the magnitude and the angle accordingly
	 * @param _b The new imaginary part
	 */
	public void setB(double _b) {
		this.b = _b;
		this.r = Math.sqrt(this.a * this.a + this.b * this.b);
		this.phi = Math.atan(this.b / this.a);
	}

	/***
	 * Modifies the magnitude of the complex number. Recalculates the real part and the imaginary part accordingly
	 * @param _r The new magnitude
	 */
	public void setR(double _r) {
		this.r = _r;
		this.a = this.r * Math.cos(this.phi);
		this.b = this.r * Math.sin(this.phi);
	}

	/***
	 * Modifies the angle of the complex number. Recalculates the real part and the imaginary part accordingly
	 * @param angle The new angle
	 */
	public void setAngle(double angle) {
		this.phi = angle;
		this.a = this.r * Math.cos(this.phi);
		this.b = this.r * Math.sin(this.phi);
	}

	// Getters

	/***
	 * @return Returns the real part of the calling Complex
	 */
	public double getA() {
		return this.a;
	}

	/***
	 * @return Returns the imaginary part of the calling Complex
	 */
	public double getB() {
		return this.b;
	}

	/***
	 * @return Returns the magnitude of the calling Complex
	 */
	public double getR() {
		return this.r;
	}

	/***
	 * @return Returns the angle of the calling Complex
	 */
	public double getPhi() {
		return this.phi;
	}

	/***
	 * @return Returns the angle of the calling Complex
	 */
	public double getAngle() {
		return this.getPhi();
	}

	/***
	 * @return Returns a string representing the Complex. Format: "<i>real part</i> + <i>imaginary part</i> i";
	 */
	public String toString() {
		return String.format("%1.2f + %1.2f i", this.a, this.b);
	}

	/***
	 * @return Returns a string representing the Complex in polar form. Format: "<i>mag</i>f * (cos(<i>angle</i>) + i * sin(<i>angle</i>))";
	 */
	public String toStringPolar() {
		return String.format("%1.2f * (cos(%f) + i * sin(%f))", this.r, this.phi, this.phi);
	}
}

